cpp/src/arrow/compute/kernels/vector_buffer.cc

```diff
@@ -0,0 +1,345 @@
+    RETURN_NOT_OK(range_lengths->Finish(&range_lengths_arr));
+    return StructArray::Make(
+        {range_starts_arr, range_offsets_arr, range_lengths_arr},
+        {field("start", uint64()), field("offset", uint64()), field("length", uint64())});
+  }
+
+  static Result<std::shared_ptr<Array>> Exec(const std::shared_ptr<ArrayData>& input) {
+    UInt64Builder range_starts, range_offsets, range_lengths;
+    GetByteRangesArray self{input,         input->offset,  input->length,
+                            &range_starts, &range_offsets, &range_lengths};
+    RETURN_NOT_OK(VisitTypeInline(*input->type, &self));
+    return self.MakeRanges();
+  }
+};
+
+const FunctionDoc byte_ranges_doc(
+    "Returns an array of byte ranges referenced by the input",
+    (R"(The output is a StructArray {"start": int64, "offset": int64, "length":)"
+     " int64} where each item represents a range of memory addressed by buffers in"
+     " the input.  The ranges should have no overlap even if buffers are shared in the"
+     " input.  If any of the arrays are sliced zero-copy views of the data this method"
+     " will return the sliced ranges that are referenced.  There is a slight exception"
+     " in the case of dictionary arrays.  If a dictionary array is sliced we will still"
+     " assume the entire dictionary is referenced by the sliced offsets array.  This"
+     " will lead to overestimation if not all dictionary values are referenced.  If"
+     " the first argument is a scalar this will return an empty array since there are no"
+     " buffers referenced by the scalar."),
+    {"input"});
+
+class ByteRangesMetaFunction : public MetaFunction {
+ public:
+  ByteRangesMetaFunction()
+      : MetaFunction("byte_ranges", Arity::Unary(), &byte_ranges_doc) {}
+
+  Result<Datum> ExecuteImpl(const std::vector<Datum>& args,
+                            const FunctionOptions* options,
+                            ExecContext* ctx) const override {
+    DCHECK_GT(args.size(), 0);
+    switch (args[0].kind()) {
+      case Datum::SCALAR: {
+        std::unique_ptr<ArrayBuilder> builder;
+        RETURN_NOT_OK(
+            MakeBuilder(ctx->memory_pool(), GetByteRangesArray::RangesType(), &builder));
+        RETURN_NOT_OK(builder->Resize(0));
+        return builder->Finish();
+      }
+      case Datum::ARRAY: {
+        const std::shared_ptr<ArrayData>& array = args[0].array();
+        DCHECK(array);
+        return GetByteRangesArray::Exec(array);
+      }
+      case Datum::CHUNKED_ARRAY: {
+        const std::shared_ptr<ChunkedArray>& chunked_array = args[0].chunked_array();
+        DCHECK(chunked_array);
+        ArrayVector chunks;
+        for (const auto& chunk : chunked_array->chunks()) {
+          ARROW_ASSIGN_OR_RAISE(std::shared_ptr<Array> chunk_ranges,
+                                GetByteRangesArray::Exec(chunk->data()));
+          chunks.push_back(chunk_ranges);
+        }
+        return Concatenate(chunks, ctx->memory_pool());
```

Comment:
I moved this out of the compute layer and changed the functions to return lengths instead of arrays.  If someone needs all referenced ranges of a table they can cross that bridge when they come to it.